toolchain/sem_ir/constant.h

```diff
@@ -32,15 +32,51 @@ enum class ConstantDependence : uint8_t {
 
 // Information about a symbolic constant value. These are indexed by
 // `ConstantId`s for which `is_symbolic` is true.
+//
+// A constant value is represented by a fully-evaluated inst, called a "constant
+// inst", which may depend on other constant insts. That constant inst fully
+// represents the constant value in itself, but for symbolic constant values we
+// sometimes need efficient access to metadata about the mapping between the
+// constant and corresponding constants in specifics of its enclosing generic.
+// As a result, the ID of a concrete constant directly encodes the ID of the
+// constant inst, but the ID of a symbolic constant is an index into a table of
+// `SymbolicConstant` entries containing that metadata, as well as the constant
+// inst ID.
+//
+// The price of this optimization is that the constant value's ID depends on the
+// enclosing generic, which isn't semantically relevant unless we're
+// specifically operating on the generic -> specific mapping. As a result, every
+// symbolic constant is represented by two `SymbolicConstant`s, with separate
+// IDs: one with that additional metadata, and one without it. The form with
+// additional metadata is called an "attached constant", and the form without it
+// is an "unattached constant". Note that constants in separate generics may be
+// represented by the same unattached constant. In general, only one of these
+// IDs is correct to use in a given situation; `ConstantValueStore` can be used
+// to map between them if necessary.
+//
+// Equivalently, you can think of an unattached constant as being implicitly
+// parameterized by the `bind_symbolic_name` constant insts that it depends on,
+// whereas an attached constant explicitly binds them to parameters of the
+// enclosing generic. It's the difference between "`Vector(T)` where `T` is some
+// value of type `type`" and "`Vector(T)` where `T` is the `T:! type` parameter
+// of this particular enclosing generic".
+//
+// TODO: consider instead keeping this metadata in a separate hash map keyed by
+// a `GenericId`/`ConstantId` pair, so that each constant has a single
+// `ConstantId`, rather than separate attached and unattached IDs.
 struct SymbolicConstant : Printable<SymbolicConstant> {
-  // The constant instruction that defines the value of this symbolic constant.
```

Comment:
We also use the "defines the value of a constant" terminology [in the definition of the `InstConstantKind` values](https://github.com/carbon-language/carbon-lang/blob/dc61460b5aec137295c4f3bc410f37f12f945f0e/toolchain/sem_ir/inst_kind.h#L35), so if you're changing it here, can you change it there too?

Reply:
Done. Note that I made a couple of drive-by fixes while I was there.